src/auth.rs

```diff
@@ -0,0 +1,78 @@
+//! Auth implementation for bore client and server.
+
+use anyhow::{bail, Result};
+use chacha20poly1305::aead::{AeadInPlace, NewAead};
+use chacha20poly1305::{ChaCha20Poly1305, Nonce};
+use rand::RngCore;
+
+/// Length of the secret key used for encrypted secret transmission.
+const KEY_LEN: usize = 32;
+
+/// Length of the nonce used for encrypted secret transmission.
+const NONCE_LEN: usize = 12;
+
+/// Secret key used for encryption.
+pub type Key = (ChaCha20Poly1305, String);
+
+/// Generate a key from a secret.
+pub fn key_from_sec(sec: &str) -> Result<Key> {
+    let sec_p = sec_padded(sec);
+    let sec_bytes = sec.as_bytes();
+    if sec_bytes.len() > KEY_LEN {
+        bail!("secret must be 32 bytes or fewer");
+    }
```

Comment:
Is it possible to just pass the secret through a SHA-2 digest, as the seed of an PRNG, or some other one-way key generation approach that removes the 32 byte restriction? That might be a slightly nicer interface.